lib/auth/azure_certs.go

```diff
@@ -0,0 +1,247 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"net/http"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+var allowedAzureCommonNames = []string{
+	"metadata.azure.com",
+	"metadata.azure.us",
+	"metadata.azure.cn",
+	"metadata.microsoftazure.de",
+}
+
+func isAllowedDomain(rawURL string, domains []string) bool {
+	addr, err := utils.ParseAddr(rawURL)
+	if err != nil {
+		return false
+	}
+	hostname := addr.Host()
+	for _, domain := range domains {
+		if hostname == domain || strings.HasSuffix(hostname, "."+domain) {
+			return true
+		}
+	}
+	return false
+}
+
+// getAzureIssuerCert fetches a x509 certificate's issuing certificate.
+func getAzureIssuerCert(ctx context.Context, cert *x509.Certificate) (*x509.Certificate, error) {
+	if len(cert.IssuingCertificateURL) == 0 {
+		return nil, nil
+	}
+
+	httpClient, err := defaults.HTTPClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Azure sends only one issuing cert.
+	issuerURL := cert.IssuingCertificateURL[0]
+	if !isAllowedDomain(issuerURL, allowedAzureCommonNames) {
+		return nil, trace.AccessDenied("certificate issued by an unexpected authority")
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, issuerURL, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+	body, err := utils.ReadAtMost(resp.Body, teleport.MaxHTTPResponseSize)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, trace.ReadError(resp.StatusCode, body)
+	}
+
+	issuerCert, err := x509.ParseCertificate(body)
+	return issuerCert, trace.Wrap(err)
+}
+
+func getAzureRootCerts() ([]*x509.Certificate, error) {
+	var certs []*x509.Certificate
+	for _, certBytes := range rawAzureCerts {
+		certPEM, _ := pem.Decode(certBytes)
+		cert, err := x509.ParseCertificate(certPEM.Bytes)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		certs = append(certs, cert)
+	}
+	return certs, nil
+}
+
+// These certificates can be found at
+// https://learn.microsoft.com/en-us/azure/security/fundamentals/azure-ca-details.
+// Pinned on 2023-01-25.
+var rawAzureCerts = [][]byte{
+	// Baltimore CyberTrust Root
+	[]byte(`-----BEGIN CERTIFICATE-----
```

Comment:
Perhaps not an issue for now, but I wonder if theres anything we can do to identify easily when a new certificate authority is announced ?